openrouteservice/src/main/java/org/heigit/ors/matrix/algorithms/core/CoreMatrixAlgorithm.java

```diff
@@ -0,0 +1,548 @@
+import org.heigit.ors.routing.graphhopper.extensions.core.CoreMatrixFilter;
+import org.heigit.ors.routing.graphhopper.extensions.edgefilters.EdgeFilterSequence;
+import org.heigit.ors.routing.graphhopper.extensions.storages.AveragedMultiTreeSPEntry;
+import org.heigit.ors.routing.graphhopper.extensions.storages.MultiTreeSPEntryItem;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.PriorityQueue;
+
+/**
+ * A Core and Dijkstra based algorithm that calculates the weights from multiple start to multiple goal nodes.
+ * Using core and true many to many.
+ * @author Hendrik Leuschner
+ */
+public class CoreMatrixAlgorithm extends AbstractMatrixAlgorithm {
+    protected int coreNodeLevel;
+    protected int maxVisitedNodes = Integer.MAX_VALUE;
+    protected int visitedNodes;
+    private int treeEntrySize;
+    private boolean hasTurnWeighting = false;
+    private boolean swap = false;
+
+    private PriorityQueue<AveragedMultiTreeSPEntry> upwardQueue;
+    private IntHashSet coreEntryPoints;
+    private IntHashSet coreExitPoints;
+    private IntObjectMap<AveragedMultiTreeSPEntry> bestWeightMap;
+    private IntObjectMap<List<AveragedMultiTreeSPEntry>> bestWeightMapCore;
+    private IntObjectMap<AveragedMultiTreeSPEntry> targetMap;
+    private IntHashSet targetSet;
+    private MultiTreeMetricsExtractor pathMetricsExtractor;
+    private CoreDijkstraFilter additionalCoreEdgeFilter;
+    private CHGraph chGraph;
+    private SubGraph targetGraph;
+    private TurnWeighting turnWeighting;
+
+    @Override
+    public void init(MatrixRequest req, GraphHopper gh, Graph graph, FlagEncoder encoder, Weighting weighting) {
+        if (weighting instanceof TurnWeighting) {
+            hasTurnWeighting = true;
+            turnWeighting = (TurnWeighting) weighting;
+        }
+        weighting = new PreparationWeighting(weighting);
+        super.init(req, gh, graph, encoder, weighting);
+        try {
+            chGraph = graph instanceof CHGraph ? (CHGraph) graph : (CHGraph) ((QueryGraph) graph).getMainGraph();
+        } catch (ClassCastException e) {
+            throw new ClassCastException(e.getMessage());
+        }
+        coreNodeLevel = chGraph.getNodes() + 1;
+        pathMetricsExtractor = new MultiTreeMetricsExtractor(req.getMetrics(), graph, this.encoder, weighting, req.getUnits());
+        initCollections(10);
+        initFilter(null, chGraph);
+    }
+
+    public void init(MatrixRequest req, GraphHopper gh, Graph graph, FlagEncoder encoder, Weighting weighting, EdgeFilter additionalEdgeFilter) {
+        this.init(req, gh, graph, encoder, weighting);
+        initFilter(additionalEdgeFilter, chGraph);
+    }
+
+    public void init(MatrixRequest req, Graph graph, FlagEncoder encoder, Weighting weighting, EdgeFilter additionalEdgeFilter) {
+        this.init(req, null, graph, encoder, weighting);
+        initFilter(additionalEdgeFilter, chGraph);
```

Comment:
In here `initFilter` gets called for the third time.

Reply:
You're right, one of them can be removed. Those are different init functions under different circumstances (1. to adhere to `AbstractMatrixAlgorithm`, then to init with a filter, then to init without a GH instance). One of the initFilters has been removed.